src/main/java/io/cdap/plugin/github/source/batch/GithubInputFormat.java

```diff
@@ -0,0 +1,110 @@
+import com.google.gson.GsonBuilder;
+import io.cdap.plugin.github.source.common.GitHubRequestFactory;
+import org.apache.hadoop.conf.Configuration;
+import org.apache.hadoop.mapreduce.InputFormat;
+import org.apache.hadoop.mapreduce.InputSplit;
+import org.apache.hadoop.mapreduce.JobContext;
+import org.apache.hadoop.mapreduce.RecordReader;
+import org.apache.hadoop.mapreduce.TaskAttemptContext;
+
+import java.io.IOException;
+import java.util.Collections;
+import java.util.List;
+import java.util.Objects;
+import java.util.stream.Collectors;
+import java.util.stream.IntStream;
+
+/**
+ * InputFormat for mapreduce job, which provides a single split of data.
+ */
+public class GithubInputFormat extends InputFormat {
+
+  private static final Gson GSON = new GsonBuilder().create();
+
+  @Override
+  public List<InputSplit> getSplits(JobContext jobContext) throws IOException {
+    Configuration conf = jobContext.getConfiguration();
+    String configJson = conf.get(GithubFormatProvider.PROPERTY_CONFIG_JSON);
+    GithubBatchSourceConfig config = GSON.fromJson(configJson, GithubBatchSourceConfig.class);
+
+    String url = GitHubRequestFactory.generateFirstCallUrl(config);
+    HttpRequest httpRequest = GitHubRequestFactory.buildRequest(url, config.getAuthorizationToken());
+    HttpResponse response = httpRequest.execute();
+    Object paginationMetadata = response.getHeaders().get("Link");
+    if (Objects.nonNull(paginationMetadata)) {
+      String paginationUrls = (String) ((List<?>) paginationMetadata).get(0);
+      String lastLink = getLastLink(paginationUrls);
+
+      Integer totalPagesCount = getTotalPagesCount(lastLink);
+      return IntStream.range(1, totalPagesCount)
+        .mapToObj(pageNumber -> {
+          String pageLink = transformLink(lastLink, pageNumber);
+          return new GithubSplit(pageLink);
+        })
+        .collect(Collectors.toList());
+    } else {
+      return Collections.singletonList(new GithubSplit(url));
+    }
+  }
+
+  @Override
+  public RecordReader createRecordReader(InputSplit inputSplit, TaskAttemptContext taskAttemptContext) {
+    Configuration conf = taskAttemptContext.getConfiguration();
+    String configJson = conf.get(GithubFormatProvider.PROPERTY_CONFIG_JSON);
+    GithubBatchSourceConfig config = GSON.fromJson(configJson, GithubBatchSourceConfig.class);
+    return new GithubRecordReader(config, ((GithubSplit) inputSplit).getLink());
+  }
+
+  private String transformLink(String link, Integer pageNumber) {
+    int indexOfLastParamValue = link.lastIndexOf("=");
+    String substring = link.substring(0, indexOfLastParamValue + 1);
+    return substring + pageNumber;
+  }
+
+  private String getLastLink(String paginationHeader) {
+    String[] links = paginationHeader.split(",");
+    for (String link : links) {
+      if (link.contains("last")) {
+        String[] parts = link.split(";");
+        String url = parts[0];
+        return url.substring(2, url.length() - 1);
```

Comment:
Please add a comment that describes what `getLastLink` does. What does paginationHeader look like?

Reply:
https://issues.cask.co/browse/PLUGIN-384